src/daft-warc/src/lib.rs

```diff
@@ -0,0 +1,575 @@
+    }
+
+    async fn read_chunk(&mut self) -> DaftResult<Option<RecordBatch>> {
+        let mut line_buf = Vec::with_capacity(4096);
+
+        loop {
+            if self.unprocessed_record {
+                return Ok(Some(self.process_arrays()?));
+            }
+
+            line_buf.clear();
+            match self.reader.read_until(b'\n', &mut line_buf).await {
+                Ok(0) => {
+                    return if self.unprocessed_record {
+                        self.unprocessed_record = false;
+                        Ok(Some(self.process_arrays()?))
+                    } else {
+                        Ok(None)
+                    };
+                }
+                Ok(2)
+                    if line_buf[0] == b'\r'
+                        && line_buf[1] == b'\n'
+                        && self.header_state.content_length.is_some() =>
+                {
+                    self.bytes_read += 2;
+
+                    // Create properly escaped JSON object from accumulated headers
+                    let header_json = if self.header_state.header_lines.is_empty() {
+                        "{}".to_string()
+                    } else {
+                        let mut obj = serde_json::Map::new();
+                        for (key, value) in &self.header_state.header_lines {
+                            obj.insert(key.clone(), serde_json::Value::String(value.clone()));
+                        }
+                        serde_json::to_string(&obj).unwrap_or_else(|_| "{}".to_string())
+                    };
+                    self.header_array.push(Some(header_json));
+
+                    let len = self
+                        .header_state
+                        .content_length
+                        .expect("Content length is required");
+
+                    if len == 0 {
+                        self.content_array.push_null();
+                    } else {
+                        let mut content_vec = vec![0; len];
+                        self.reader.read_exact(&mut content_vec).await?;
+                        self.content_array.push(Some(content_vec));
+                        self.bytes_read += len;
+                    }
+
+                    self.record_id_array
+                        .push(self.header_state.record_id.map(|id| id.to_string()));
+                    self.warc_type_array
+                        .push(self.header_state.warc_type.take().map(|t| t.to_string()));
+                    self.warc_date_array.push(
+                        self.header_state
+                            .warc_date
+                            .and_then(|d| d.timestamp_nanos_opt()),
+                    );
+                    self.warc_content_length_array
+                        .push(self.header_state.content_length.map(|len| len as i64));
+
+                    self.unprocessed_record = true;
+                    self.header_state.reset();
+                    if self.bytes_read >= self.chunk_size {
```

Comment:
I think this chunk size refers to rows rather than bytes. I think this should `self.rows_read >= self.chunk_size`